File: string_search.js
function isValid(s) {
    let flag = true;
    let str_list = [];
    let str_map = {')':'(', ']':'[', '}':'{'};

    for (let char of s) {
        if (Object.values(str_map).includes(char)) {
            str_list.push(char);
        } else if (Object.keys(str_map).includes(char)) {
            if ((str_list.length === 0) || (str_map[char] !== str_list.pop())) {
                return !flag;
            }
        }
    }
    return flag;
}

// テストする文字列定義
let test_str = ['()', '([]){}', '({)}'];

for (let char of test_str) {
    console.log(`入力文字列：${char}`);
    console.log(`判定結果　：${isValid(char)}`);
}

File: string_search.rb
def is_valid(s)
    flag = true
    str_list = []
    str_map = {')' => '(', ']' => '[', '}' => '{'}

    s.each_char do |char|
        if str_map.values.include?(char)
            str_list.push(char)
        elsif str_map.keys.include?(char)
            if str_list.empty? || str_map[char] != str_list.pop
                return !flag
            end
        end
    end

    flag
end

# テストする文字列定義
test_str = ['()', '([]){}', '({)}']

test_str.each do |char|
    puts "入力文字列：#{char}"
    puts "判定結果　：#{is_valid(char)}"
end